Runtime/Data/Operation/Cache/ValueCache.cs

```diff
@@ -0,0 +1,68 @@
+﻿namespace Zinnia.Data.Operation.Cache
+{
+    using Malimbe.BehaviourStateRequirementMethod;
+    using UnityEngine;
+    using UnityEngine.Events;
+
+    /// <summary>
+    /// Caches <see cref="TValue"/> data and invokves an appropriate event when the cache is updated.
+    /// </summary>
+    /// <typeparam name="TValue">The type of data to cache.</typeparam>
+    /// <typeparam name="TEvent">The event to invoke.</typeparam>
+    public abstract class ValueCache<TValue, TEvent> : MonoBehaviour where TEvent : UnityEvent<TValue>, new()
+    {
+        /// <summary>
+        /// Emitted when the cached data is updated and has been modified to a new value.
+        /// </summary>
+        public TEvent Modified = new TEvent();
+        /// <summary>
+        /// Emitted when the cached data is updated but the value is unmodified.
+        /// </summary>
+        public TEvent Unmodified = new TEvent();
+
+        /// <summary>
+        /// The cache of data.
+        /// </summary>
+        public TValue Data
+        {
+            get
+            {
+                return data;
+            }
+            [RequiresBehaviourState]
+            set
+            {
+                if (AreEqual(value, data))
+                {
+                    Unmodified?.Invoke(value);
+                }
+                else
+                {
+                    Modified?.Invoke(value);
+                }
+                data = value;
+            }
+        }
+        /// <summary>
+        /// The backing field for <see cref="Data"/>.
+        /// </summary>
+        private TValue data;
+
+        /// <summary>
+        /// Clears the cache by setting it to the default value.
+        /// </summary>
+        [RequiresBehaviourState]
+        public virtual void ClearCache()
+        {
+            Data = default;
+        }
+
+        /// <summary>
+        /// Determines if the two given values are considered equal.
+        /// </summary>
+        /// <param name="a">The first value to check.</param>
+        /// <param name="b">The second value to check.</param>
+        /// <returns>Whether the two given values are considered equal.</returns>
+        protected abstract bool AreEqual(TValue a, TValue b);
```

Comment:
This could be virtual and offer a base implementation by returning `EqualityComparer<TValue>.Default.Equals(a, b)` I think. Worth it? As far as I can see this change would then remove the need to have subclasses implement this method for the `int`, `string` and `GameObject` cases at least (the `float` one wants the custom approximate check, and there might be others I didn't check).